### DISCUSSION TITLE
How to test Non blocking Spring Kafka Retry using Spring boot embedded kafka broker?

### DISCUSSION THREAD

**Discussion by hariraogotit** [spring-projects/spring-kafka#2755] [Q&A]
Team,

  I have been using WaitForCompletionMessagePublisher for testing normal kafka messages 

```java
 waitForCompletionMessagePublisher.sendToTopicAndWaitUntilProcessedWithoutException(
        proto.toByteArray(), topic, registry, kafkaTemplate, partitions, 2000L);
``` 

But how to test Spring kafka non blocking retry? I did a combination of the following but it does not wait for the non blocking retry to happen?

```java
doThrow(new RuntimeException("Failing this,to validate Kafka retry"))
            .when(customerManagerClient)
            .getCust(any());

waitForCompletionMessagePublisher.sendToTopicAndWaitUntilProcessedWithException( // with exception btw
            proto.toByteArray(), topic, registry, kafkaTemplate, partitions, 2000L);

verify(customerManagerClient,
                    atLeast(3))
            .getCust(any());


``` 

Any leads are much appreciated. Thanks

**Comment by garyrussell** [comment]
I don't know what `WaitForCompletionMessagePublisher` is. You need to provide much more information, configuration, code.

**Reply by hariraogotit** [reply]
Apologies @garyrussell 

I just realised its a proprietary code so I will not be able to share which means you will not be able to help. I assumed WaitForCompletionMessagePublisher belonged to Spring.

So will debug further and for now will close this request. 

Thanks for coming back to me.
